src/Native/gc/sample/gcenv.windows.cpp

```diff
@@ -0,0 +1,444 @@
+//  ms - pointer to the structure that will be filled in with the memory status
+void GCToOSInterface::GetMemoryStatus(GCMemoryStatus* ms)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+    }
+    CONTRACTL_END;
+
+    MEMORYSTATUSEX memStatus;
+
+    memStatus.dwLength = sizeof(MEMORYSTATUSEX);
+    BOOL fRet = GlobalMemoryStatusEx(&memStatus);
+    _ASSERTE (fRet);
+
+    // If the machine has more RAM than virtual address limit, let us cap it.
+    // Our GC can never use more than virtual address limit.
+    if (memStatus.ullAvailPhys > memStatus.ullTotalVirtual)
+    {
+        memStatus.ullAvailPhys = memStatus.ullAvailVirtual;
+    }
+
+    // Convert Windows struct to abstract struct
+    ms->dwMemoryLoad              = memStatus.dwMemoryLoad           ;
+    ms->ullTotalPhys              = memStatus.ullTotalPhys           ;
+    ms->ullAvailPhys              = memStatus.ullAvailPhys           ;
+    ms->ullTotalPageFile          = memStatus.ullTotalPageFile       ;
+    ms->ullAvailPageFile          = memStatus.ullAvailPageFile       ;
+    ms->ullTotalVirtual           = memStatus.ullTotalVirtual        ;
+    ms->ullAvailVirtual           = memStatus.ullAvailVirtual        ;
+}
+
+// Get a high precision performance counter
+// Return:
+//  The counter value
+int64_t GCToOSInterface::QueryPerformanceCounter()
+{
+    LARGE_INTEGER ts;
+    if (!::QueryPerformanceCounter(&ts))
+    {
+        _ASSERTE(!"Fatal Error - cannot query performance counter.");
+        abort();
+    }
+
+    return ts.QuadPart;
+}
+
+// Get a frequency of the high precision performance counter
+// Return:
+//  The counter frequency
+int64_t GCToOSInterface::QueryPerformanceFrequency()
+{
+    LARGE_INTEGER frequency;
+    if (!::QueryPerformanceFrequency(&frequency))
+    {
+        _ASSERTE(!"Fatal Error - cannot query performance counter.");
+        abort();
+    }
+
+    return frequency.QuadPart;
+}
+
+// Get a time stamp with a low precision
+// Return:
+//  Time stamp in milliseconds
+uint32_t GCToOSInterface::GetLowPrecisionTimeStamp()
+{
+    return ::GetTickCount();
+}
+
+// Parameters of the GC thread stub
+struct GCThreadStubParam
+{
+    GCThreadFunction GCThreadFunction;
+    void* GCThreadParam;
+};
+
+// GC thread stub to convert GC thread function to an OS specific thread function
+static DWORD GCThreadStub(void* param)
+{
+    GCThreadStubParam *stubParam = (GCThreadStubParam*)param;
+    stubParam->GCThreadFunction(stubParam->GCThreadParam);
+
+    return 0;
+}
+
+// Create a new thread
+// Parameters:
+//  function - the function to be executed by the thread
+//  param    - parameters of the thread
+//  affinity - processor affinity of the thread
+// Return:
+//  true if it has succeeded, false if it has failed
+bool GCToOSInterface::CreateThread(GCThreadFunction function, void* param, GCThreadAffinity* affinity)
+{
+    DWORD thread_id;
+    GCThreadStubParam stubParam;
+
+    stubParam.GCThreadFunction = function;
```

Comment:
Maybe pick up the few GC changes that were made in CoreCLR over the last couple of weeks?